R/code-generation.R

```diff
@@ -0,0 +1,31 @@
+dm_f_list <- function(f_list = list()) {
+  structure(f_list, class = c("dm_f_list", class(f_list)))
+}
+
+add_call <- function(f_list, fn_call) {
+  fn <- function(.) NULL
+  body(fn) <- enexpr(fn_call)
+  dm_f_list(append(f_list, fn))
+}
+
+#' @export
+print.dm_f_list <- function(x, ...) {
+  show_dm_f_list(x)
+}
+
+show_dm_f_list <- function(x) {
+  if (is_empty(x)) {
+    invisible(x)
+  } else {
+    purrr::map_chr(x, ~ body(.) %>% call_to_char()) %>%
+      paste0("  ", ., collapse = " %>%\n") %>%
+      cat()
+    invisible(x)
+  }
+}
+
+call_to_char <- function(body) {
+  deparse(body) %>%
+    paste0(collapse = "") %>%
+    gsub("\\s+", " ", .)
```

Comment:
Good points. I am wondering though, how to produce nicely formatted code, e.g.:
``` r
call_to_char <- function(body) {
  paste0(trimws(deparse(body)), collapse = "\n  ")
}

fn_body <- rlang::expr(dm_add_fk(., table = flights, columns = tailnum, ref_table = airlines, check = TRUE, on_delete = "no_action"))
# Expression:
fn_body
#> dm_add_fk(., table = flights, columns = tailnum, ref_table = airlines, 
#>     check = TRUE, on_delete = "no_action")

# Character:
call_to_char(fn_body)
#> [1] "dm_add_fk(., table = flights, columns = tailnum, ref_table = airlines,\n  check = TRUE, on_delete = \"no_action\")"

# finally:
cat(call_to_char(fn_body))
#> dm_add_fk(., table = flights, columns = tailnum, ref_table = airlines,
#>   check = TRUE, on_delete = "no_action")
```

<sup>Created on 2022-06-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>
shall we insert a "\n  " after each comma, in case the total width exceeds a limit?

Reply:
Should we use {styler} right before printing ?

Reply:
Also what about using `rlang::expr_deparse()` rather than `deparse()` ? It handles some corner cases and has a default width cutoff of `106` which can be set through options. 

```
deparse(quote(x := 1))
#> [1] "`:=`(x, 1)"
rlang::expr_deparse(quote(x := 1))
#> [1] "x := 1"

deparse(quote(?x))
#> [1] "`?`(x)"
rlang::expr_deparse(quote(?x))
#> [1] "?x"
```

Reply:
Thanks! `rlang::expr_deparse()` is nice, using it.
Maybe for now I would avoid using {styler}, because it seems to me that it doesn't take care of the decision to spread code nicely over several lines if it's too wide (looked at `styler::style_text()`). Also it would be one more Import-dependency.